src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -1561,6 +1591,88 @@ void InstanceKlass::set_initialization_state_and_notify(ClassState state, TRAPS)
   }
 }
 
+void InstanceKlass::notify_strict_static_access(int field_index, bool is_writing, TRAPS) {
+  guarantee(field_index >= 0 && field_index < fields_status()->length(), "valid field index");
+  DEBUG_ONLY(FieldInfo debugfi = field(field_index));
+  assert(debugfi.access_flags().is_strict(), "");
+  assert(debugfi.access_flags().is_static(), "");
+  FieldStatus& fs = *fields_status()->adr_at(field_index);
+  LogTarget(Trace, class, init) lt;
+  if (lt.is_enabled()) {
+    ResourceMark rm(THREAD);
+    LogStream ls(lt);
+    FieldInfo fi = field(field_index);
+    ls.print("notify %s %s %s%s ",
+             external_name(), is_writing? "Write" : "Read",
+             fs.is_strict_static_unset() ? "Unset" : "(set)",
+             fs.is_strict_static_unread() ? "+Unread" : "");
+    fi.print(&ls, constants());
+  }
+  if (fs.is_strict_static_unset()) {
+    assert(fs.is_strict_static_unread(), "ClassFileParser resp.");
+    // If it is not set, there are only two reasonable things we can do here:
+    // - mark it set if this is putstatic
+    // - throw an error (Read-Before-Write) if this is getstatic
+    //
+    // A third less-reasonable thing is note an internal error if the
+    // JDK reflection logic has allowed another thread to sneak into
+    // the <clinit> critical section.
+    if (!is_reentrant_initialization(THREAD)) {
+      // The unset state is (or should be) transient, and observable only in one
+      // thread during the execution of <clinit>.  Something is wrong here, and
+      // we should just throw.
+      if (is_in_error_state()) {
+        oop init_error = get_initialization_error(THREAD);
+        if (init_error != nullptr) {
+          THROW_OOP(init_error);
+        }
+      }
+      THROW_MSG(vmSymbols::java_lang_InternalError(), "unscoped access to strict static");
+    } else if (is_writing) {
+      // clear the "unset" bit, since the field is actually going to be written
+      fs.update_strict_static_unset(false);
+    } else {
+      // throw an IllegalStateException, since we are reading before writing
+      // see also InstanceKlass::initialize_impl, Step 8 (at end)
+      Symbol* bad_strict_static = field(field_index).name(constants());
+      throw_strict_static_exception(bad_strict_static, "is unset before first read in", CHECK);
+    }
+  } else {
+    // Experimentally, enforce additional proposed conditions after the first write.
```

Comment:
```suggestion
    // Ensure no write after read for final strict statics
```
I think this comment may be left over from the multi-option initial prototype.  It might be worth going over all the comments and making sure they still apply now that we've settled on a single approach